webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/AbstractIndexDescriptor.java

```diff
@@ -43,26 +51,29 @@ public String getMappingsClasspathFilename() {
 
   @Override
   public String getAllVersionsIndexNameRegexPattern() {
-    return String.format("%s-%s-\\d.*", getPrefixAndComponentName(), getIndexName());
+    if (isGlobalPrefix) {
+      return String.format(
+          "%s%s-%s-\\d.*", formattedIndexPrefix(), getComponentName(), getIndexName());
+    } else {
+      return String.format("%s-%s-\\d.*", getIndexPrefix(), getIndexName());
+    }
   }
 
   @Override
   public String getVersion() {
     return "1.0.0";
   }
 
+  private String formattedIndexPrefix() {
+    return getIndexPrefix().isBlank() ? "" : getIndexPrefix() + "-";
+  }
+
   public String getIndexPrefix() {
     return indexPrefix;
   }
 
-  private String getPrefixAndComponentName() {
-    // Cannot start index with "-" so must not append "-" for empty prefix
-    final var prefix = getIndexPrefix().isBlank() ? getIndexPrefix() : getIndexPrefix() + "-";
-
-    // Legacy descriptors have the same index prefix as component name this avoids duplication.
-    return getIndexPrefix().contains(getComponentName())
-        ? getIndexPrefix()
-        : prefix + getComponentName();
+  public void setIsGlobalPrefix(final boolean isGlobalPrefix) {
```

Comment:
🔧 💭 A builder-style `withHasGlobalPrefix()` that returns an instance (i.e. `return `this`) could help getting rid of the list, then loop then map in the ResourceProvider class

Reply:
I thought about this there are two solutions and I think this is probably the cleanest. If I want the `withHasGlobalPrefix` to be in the `AbstractIndexDescriptor` when I `return this` it will return an instance of `AbstractIndexDescriptor` instead of the implementation type. 

Possible solutions
- Make `AbstractIndexDescriptor` a generic class such that each descriptor class is now `Descriptor extends AbstractIndexDescriptor<Descriptor>` not that clean and requires a lot of change.
- Current approach, not ideal as the descriptor is now not immutable.

If there's a better way to implement the `with` syntax other than generics I'd be open to it.